psydac/mapping/discrete.py

```diff
@@ -148,16 +148,15 @@ def build_mesh(self, refine_factor=1):
             Z component of the mesh
         """
 
-        mesh = self.space.eval_fields(*self._fields, refine_factor=refine_factor)
-
+        mesh = self.space.eval_fields(grid, *self._fields, refinement=refinement)
         if self.ldim == 2:
-            x_mesh = np.ascontiguousarray(mesh[..., 0:1])
-            y_mesh = np.ascontiguousarray(mesh[..., 1:2])
+            x_mesh = np.ascontiguousarray(np.reshape(mesh[0], newshape=(*mesh.shape[1:], 1)))
+            y_mesh = np.ascontiguousarray(np.reshape(mesh[1], newshape=(*mesh.shape[1:], 1)))
             z_mesh = np.zeros_like(x_mesh)
         elif self.ldim == 3:
-            x_mesh = np.ascontiguousarray(mesh[..., 0])
-            y_mesh = np.ascontiguousarray(mesh[..., 1])
-            z_mesh = np.ascontiguousarray(mesh[..., 2])
+            x_mesh = np.ascontiguousarray(mesh[0])
+            y_mesh = np.ascontiguousarray(mesh[1])
+            z_mesh = np.ascontiguousarray(mesh[2])
```

Comment:
I printed the flags and the slices are not said to be C-contiguous (or Fortran contiguous for that matter)

Reply:
That's quite strange. Isn't `mesh` a C-contiguous array?

Reply:
The array that the pyccel functions in `kernels.py` fill is C-contiguous, however, the shape is `(n_x1,n_x2,n_x3,n_fields)`. I felt like this wasn't very intuitive for the user (or a future developper) so I then do a `np.moveaxis(-1, 0)` on the array. That is probably what breaks the C-contiguity. I can change that if needed. I just need to swap dimensions around in `kernels.py`.

Reply:
Yes, `np.moveaxis(outfields, -1, 0)` returns a non-contiguous view of `outfields` [here](https://github.com/pyccel/psydac/blob/a1e6d5ccb7949bbc9774b56c290367654282f954/psydac/fem/tensor.py#L344). 

In practice it makes sense that `eval_fields` should return a list of C-contiguous arrays, rather than a larger non-contiguous array view. Do you see a reason why we shouldn't?

Reply:
I don't, I just happened to put the fields as the last dimension. I'll fix that

Reply:
While you write into the array your currently solution probably makes sense. But I would copy the result into separate arrays...

Reply:
Fixed in [a689cc7](https://github.com/pyccel/psydac/pull/193/commits/a689cc739c69eb78a1da1227de80faac20925f2b)

Reply:
You are right that I loose performance. then I guess I can go back to my previous algorithm but still return a proper list like I do in this commit.

Reply:
Fixed ? in [ece414e](https://github.com/pyccel/psydac/pull/193/commits/ece414ed9789c760a0f91550c1da6c44d95d9ae7)

Reply:
Changed with what we talked about in 0e65d68